stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -0,0 +1,281 @@
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NakamotoBlockProposal {
+    /// Proposed block
+    pub block: NakamotoBlock,
+    /// Identify the stacks/burnchain fork we are on
+    pub parent_consensus_hash: ConsensusHash,
+    /// These are all the microblocks that the proposed block
+    /// will confirm.
+    pub burn_tip: BurnchainHeaderHash,
+    /// This refers to the burn block that was the current tip
+    ///  at the time this proposal was constructed. In most cases,
+    ///  if this proposal is accepted, it will be "mined" in the next
+    ///  burn block.
+    pub burn_tip_height: u32,
+    /// Mainnet, Testnet, etc.
+    pub chain_id: u32,
+}
+
+impl StacksMessageCodec for NakamotoBlockProposal {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.block)?;
+        write_next(fd, &self.parent_consensus_hash)?;
+        write_next(fd, &self.burn_tip)?;
+        write_next(fd, &self.burn_tip_height)?;
+        write_next(fd, &self.chain_id)
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, CodecError> {
+        Ok(Self {
+            block: read_next(fd)?,
+            parent_consensus_hash: read_next(fd)?,
+            burn_tip: read_next(fd)?,
+            burn_tip_height: read_next(fd)?,
+            chain_id: read_next(fd)?,
+        })
+    }
+}
+
+/// This enum is used to supply a `reason_code` for validation
+///  rejection responses. This is serialized as an enum with string
+///  type (in jsonschema terminology).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ValidateRejectCode {
+    BadBlockHash,
+    BadTransaction,
+    InvalidBlock,
+    ChainstateError,
+    UnknownParent,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks should be rejected.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateReject {
+    pub reason: String,
+    pub reason_code: ValidateRejectCode,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: StacksBlock,
+    pub cost: ExecutionCost,
+    pub size: u64,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum NakamotoBlockProposalResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+#[derive(Clone, Default)]
+pub struct RPCBlockProposalRequestHandler {
+    pub block_proposal: Option<NakamotoBlockProposal>,
+}
+
+impl RPCBlockProposalRequestHandler {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Decode a bare transaction from the body
+    fn parse_posttransaction_octets(mut body: &[u8]) -> Result<NakamotoBlockProposal, Error> {
```

Comment:
"posttransaction"?

Reply:
Reupping this.  This method name is wrong.  You're not posting a transaction.

Reply:
Will fix